classes/converter.php

```diff
@@ -165,6 +165,30 @@ final public function convert_moodle_urls_to_pdf(array $urls, string $filename =
         }
     }
 
+    /**
+     * Generate the PDF content from the provided HTML content.
+     *
+     * @param string $htmlcontent The raw HTML content to be converted to PDF.
+     * @param array $options any additional options to pass to converter, valid options vary with converter
+     * instance, see relevant converter for further details.
+     * @return string The generated PDF content.
+     */
+    abstract protected function generate_pdf_content_from_html(string $htmlcontent, array $options = []): string;
+
+    /**
+     * Convert HTML to PDF.
+     *
+     * @param string $htmlcontent The HTML content.
+     * @param string $filename The file name to give file.
+     * @param array $options any additional options to pass to converter, valid options vary with converter
+     * instance, see relevant converter for further details.
+     * @return bool|\stored_file the file or false if file could not be created.
```

Comment:
But the actual signature has return type `stored_file` ?

Reply:
Oops! Updated :)

Reply:
I'm thinking that `bool` isn't actually appropriate. When I look at `create_pdf_file` I see it's calling `create_file_from_string`, then when I look at `create_file_from_string`  it doesn't appear to return false, see [here](https://github.com/moodle/moodle/blob/MOODLE_401_STABLE/lib/filestorage/file_storage.php#L1392).

Is there a situation where this can actually return false?

Reply:
I thought it was returning bool because our `create_pdf_file` method declared `bool|\stored_file` xD
I pushed another fixup commit.